mlx/transforms.cpp

```diff
@@ -19,6 +19,38 @@
 
 namespace mlx::core {
 
+namespace {
+
+static std::atomic<int> _tracing_counter = 0;
+
+// Create an InTracing object during tracing operations to signify to the rest
+// of the codebase that we are during tracing so evals should not throw away
+// the graph.
+struct InTracing {
+ public:
+  InTracing() {
+    tracing_counter++;
+  }
+  ~InTracing() {
+    tracing_counter--;
+  }
+
+  static bool currently_tracing() {
+    return tracing_counter > 0;
+  }
+
+ private:
+  static int tracing_counter;
+};
+
+int InTracing::tracing_counter{0};
+
+} // namespace
+
+bool currently_tracing() {
+  return InTracing::currently_tracing();
+}
```

Comment:
Not wedded to it, but `in_tracing()` seems like a slightly better match to `InTracing`